typescript/src/schema/proto.ts

```diff
@@ -237,12 +258,54 @@ function augmentService(
   return augmentedService;
 }
 
+function getCommentsMap(
+  fd: plugin.google.protobuf.IFileDescriptorProto
+): CommentsMap {
+  const comments: CommentsMap = {};
+  if (!fd || !fd.sourceCodeInfo || !fd.sourceCodeInfo.location) return comments;
+  const locations = fd.sourceCodeInfo.location;
+  locations.forEach(location => {
+    if (location.leadingComments !== null) {
+      // p is an array with format [f1, i1, f2, i2, ...]
+      // - f1 refers to the protobuf field tag
+      // - if field refer to by f1 is a slice, i1 refers to an element in that slice
+      // - f2 and i2 works recursively.
+      // So, [6, x] refers to the xth service defined in the file,
+      // since the field tag of Service is 6.
+      // [6, x, 2, y] refers to the yth method in that service,
+      // since the field tag of Method is 2.
+      const p = location.path!;
+      if (p.length === 2 && p[0] === 6) {
+        if (fd.service && fd.service[p[1]] && fd.service[p[1]].name) {
+          comments[fd.service[p[1]].name! + 'Service'] =
+            location.leadingComments || '';
+        }
+      } else if (p.length === 4 && p[2] === 2) {
+        if (
+          fd.service &&
+          fd.service[p[1]] &&
+          fd.service[p[1]].method &&
+          fd.service[p[1]].method![p[3]] &&
+          fd.service[p[1]].method![p[3]].name
+        ) {
```

Comment:
Hi Alex, is there a better way to check `fd.service[p[1]].method![p[3]].name` exists instead of checking every layer?  Even if I check layer by layer, ts still complains about `fd.service[p[1]].method` might be `null` or `undefined` so I put `fd.service[p[1]].method!` below to make it pass. But really wondering why it fails type checking. Thanks!

Reply:
It will appear soon with [optional chaining](https://github.com/microsoft/TypeScript/issues/30167) but not right now. Let's leave it as here for now, and we'll be able to refactor this when this TypeScript feature lands.

Also, the type checks are static, they cannot apply to the random array element like `p[1]`. If you saved each individual part to a variable, it would've worked (e.g. `const method = fd.service[p[1]].method; if (method && method[p[3]]) { const whatever = method[p[3]]; if (whatever) { const name = whatever.name; }` but that does not make much sense to me.

Other thing is... can we change `1` and `3` to constants if it makes sense?

Reply:
After reading the code, maybe we don't really change numbers to constants here. Just as I wrote in another comment, let's move all this stuff to a separate `comments.ts` and leave it there as it is.

Reply:
Oh I did not know we have this new `optional chaining `feature coming, sounds awesome! 
and sure thing, I will move comment related stuff to a separate ts which makes more sense and also make the code cleaner.